REGRESSION(r294536): Drawing the ImageBuffer should not invalidate its cached copied images

shallawa opened the pull request:
#### c7880bea0dcacd97dba2d8a7265f70509dd87d04
<pre>
REGRESSION(r294536): Drawing the ImageBuffer should not invalidate its cached copied images
<a href="https://bugs.webkit.org/show_bug.cgi?id=242879">https://bugs.webkit.org/show_bug.cgi?id=242879</a>
rdar://93775177

Reviewed by Simon Fraser.

IOSurface caches the CGImage which CGIOSurfaceContextCreateImage() returns. The
same cached CGImage can be returned if the backing store did not change. The
memory of the cached CGImage is mapped to the IOSurface backing store as copy-on-
write. Accessing the pixels of the IOSurface directly does not dirty the cached
CGImage. So these ImageBuffer scenarios need special handling:

Before we putPixelBuffer() to the IOSurface, we need to detach the IOSurface
cached CGImage such that its memory is copied before the pixels of the IOSurface
get changed. So we need to call invalidateCachedNativeImage() which makes CG
draws an empty rectangle. This will kick off the copy-on-write operation.

When drawing the ImageBuffer to a destination GraphicContext, a new CGImage is
created from the IOSurface or the cached CGImage is used. It does not really matter
what the state of this CGImage. CG will behave correctly as long as the CGImage
is used in its drawing pipeline. So we do not need to invoke the empty rectangle
draw hack in subsequent calls if ImageBuffer::draw() is called.

The fix is to introduce copyNativeImageForDrawing() which copy the IOSurface to
a CGImage but will not require the empty rectangle draw hack in subsequent calls.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::copyNativeImageForDrawing const):
(WebCore::ImageBuffer::draw):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::copyNativeImageForDrawing const):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage const):
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImageIfNeeded const):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing const):
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::copyNativeImage const):
(WebKit::RemoteImageBufferProxy::copyNativeImageForDrawing const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:

Canonical link: <a href="https://commits.webkit.org/252813@main">https://commits.webkit.org/252813@main</a>
</pre>

mattwoodrow commented on Source/WebCore/platform/graphics/CopyImageOptions.h:
```diff
@@ -37,4 +37,9 @@ enum class ShouldInterpolate : bool {
 };
 
+enum class VolatileCopy : bool {
+    Yes, // Copy should be destroyed before the subsequent draws.
```
So if the caller forgets to destroy the copy, we get a weird bug where the surfaces appear to magically change under us?

Is it possible to implement this using `CFGetRetainCount` to determine if we need copy-on-write behaviour?

shallawa replied:
This is a great idea.

And it tried doing it by making ImageBufferIOSurfaceBackend be a NativeImage::Observer. And I made it keep a list of the copied CGImages. And instead of checking whether copyNativeImage() was ever called since the last time the image cache was cleared, I was checking whether this list is not empty. This seems to work except for one this. NativeImage has to be destroyed on the main thread but in GPUP they are copied from the IOSurface on RemoteRenderingBackend work queue.

When running the layout tests I hit many crashes which have the same scenario. On the RemoteRenderingBackend work queue, the IOSurface backend is trying to detach itself from all the copied NativeImages because the backing store was invalidated. And on the main thread the NativeImage is being destroyed and its destructor is trying to detach itself from the IOSurface backend.

I think it is still doable and maybe we want to remove WTF::DestructionThread::Main from the NativeImage. But I think this should be addressed in a separate patch.

cdumez replied:
Ideally, "Yes" wouldn't come first in the enum. It is weird for Yes to have 0 as value. That's why we usually use { No, Yes } order in such enum classes.

smfr commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -163,6 +163,8 @@ class ImageBuffer : public ThreadSafeRefCounted<ImageBuffer, WTF::DestructionThr
     void setBackendInfo(ImageBufferBackend::Info&& parameters) { m_backendInfo = WTFMove(parameters); }
 
     WEBCORE_EXPORT virtual RefPtr<NativeImage> copyNativeImage(BackingStoreCopy = CopyBackingStore) const;
+    WEBCORE_EXPORT virtual RefPtr<NativeImage> copyVolatileNativeImage(BackingStoreCopy = CopyBackingStore) const;
```
It's odd for these to be here on ImageBuffer. As a caller, I would not know how to pick which function to call. What's "volatile" from the perspective of the caller?

If we need to expose something here, I'd suggest an OptionSet<> or enum that communicates what we need to know (like, "this is a short-lived image" or whatever). Or we change the API like we did with `sinkIntoImage`.

smfr replied:
These can be private, which avoids polluting the API. And I think they should use names like `copyNativeImageFor` or be collapsed into one function which takes an enum `Purpose` argument.

smfr commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
```diff
@@ -80,10 +82,12 @@ class WEBCORE_EXPORT ImageBufferIOSurfaceBackend : public ImageBufferCGBackend {
 
     std::unique_ptr<IOSurface> m_surface;
-    mutable bool m_mayHaveOutstandingBackingStoreReferences { false };
+    mutable bool m_mayNativeImageBeCached { false };
```
I think this should be named more like m_mayBeUsedBy.... or m_mayBeReferencedBy... The CGImageRef references the IOSurface, rather than the IOSurface "having" an image.

smfr replied:
And maybe an OptionSet<> would make this more clear.

smfr replied:
Or maybe we turn this around and use "needs" terminology: describe what needs to happen in some future scenario.

smfr commented on Source/WebCore/platform/graphics/CopyImageOptions.h:
```diff
@@ -37,4 +37,10 @@ enum class ShouldInterpolate : bool {
+enum class CopyImageOptions : uint8_t {
+    CopyBackingStore    = 1 << 0, // Guarantee subsequent draws don't affect the copy.
+    CopyVolatile        = 1 << 1, // Copy should be destroyed before the subsequent draws.
+    CopyForPixels       = 1 << 2, //
```
Maybe CopyForEncoding

shallawa committed:
REGRESSION(r294536): Drawing the ImageBuffer should not invalidate its cached copied images
https://bugs.webkit.org/show_bug.cgi?id=242879
rdar://93775177

Reviewed by Simon Fraser.

IOSurface caches the CGImage which CGIOSurfaceContextCreateImage() returns. The
same cached CGImage can be returned if the backing store did not change. The
memory of the cached CGImage is mapped to the IOSurface backing store as copy-on-
write. Accessing the pixels of the IOSurface directly does not dirty the cached
CGImage. So these ImageBuffer scenarios need special handling:

Before we putPixelBuffer() to the IOSurface, we need to detach the IOSurface
cached CGImage such that its memory is copied before the pixels of the IOSurface
get changed. So we need to call invalidateCachedNativeImage() which makes CG
draws an empty rectangle. This will kick off the copy-on-write operation.

When drawing the ImageBuffer to a destination GraphicContext, a new CGImage is
created from the IOSurface or the cached CGImage is used. It does not really matter
what the state of this CGImage. CG will behave correctly as long as the CGImage
is used in its drawing pipeline. So we do not need to invoke the empty rectangle
draw hack in subsequent calls if ImageBuffer::draw() is called.

The fix is to introduce copyNativeImageForDrawing() which copy the IOSurface to
a CGImage but will not require the empty rectangle draw hack in subsequent calls.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::copyNativeImageForDrawing const):
(WebCore::ImageBuffer::draw):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::copyNativeImageForDrawing const):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage const):
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImageIfNeeded const):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing const):
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::copyNativeImage const):
(WebKit::RemoteImageBufferProxy::copyNativeImageForDrawing const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:

Canonical link: https://commits.webkit.org/252813@main
